File: test/analyze.test.ts
import assert from 'node:assert';
import { describe, it } from 'node:test';
import { analyzeTranslations, CHECK_CODE, createAnalyzer } from '../dist/index.js';

describe('analyzeTranslations', () => {
	it('flattens nested translations before comparing them', () => {
		const result = analyzeTranslations(
			{
				locales: {
					en: { desc: { hello: 'Hello', bye: 'Goodbye' } },
					ko: { desc: { hello: '안녕하세요' } }
				}
			},
			{ target: 'en' }
		);

		assert.strictEqual(result.keyCount, 2);
		assert.strictEqual(result.success, false);
		assert.strictEqual(result.issuesByCode.NO_KEY?.[0].key, 'desc.bye');
	});

	it('uses already flattened translations as they are', () => {
		const result = analyzeTranslations(
			{ locales: { en: { 'a.b': 'Hi' }, ko: { 'a.b': '' } } },
			{ target: 'en', flattened: true }
		);

		assert.strictEqual(result.keyCount, 1);
		assert.strictEqual(result.issuesByCode.EMPTY_VALUE?.length, 1);
	});

	it('does no file system work, so it reports no files', () => {
		const result = analyzeTranslations(
			{ locales: { en: { a: 'A' }, ko: { a: 'ㄱ' } } },
			{ target: 'en' }
		);

		assert.deepStrictEqual(result.files, []);
		assert.strictEqual(result.fileFormat, null);
	});

	it('compares each group on its own', () => {
		const result = analyzeTranslations(
			{
				groups: {
					'common.json': { en: { a: 'A' }, ko: { a: 'ㄱ' } },
					'errors.json': { en: { b: 'B' }, ko: {} }
				}
			},
			{ target: 'en' }
		);

		assert.deepStrictEqual(result.groups, ['common.json', 'errors.json']);
		assert.strictEqual(result.issuesByCode.NO_KEY?.length, 1);
		assert.strictEqual(result.issuesByCode.NO_KEY?.[0].group, 'errors.json');
	});

	it('honours custom interpolation delimiters', () => {
		const input = { locales: { en: { a: 'Hello {{name}}' }, ko: { a: '안녕하세요' } } };

		assert.strictEqual(
			analyzeTranslations(input, {
				target: 'en',
				interpolationPrefix: '{{',
				interpolationSuffix: '}}'
			}).issuesByCode.NO_INTERPOLATION_KEY?.length,
			1
		);

		// The default single-brace delimiters do not recognise `{{name}}`, so the
		// missing placeholder goes unnoticed.
		assert.strictEqual(
			analyzeTranslations(input, { target: 'en' }).issuesByCode.NO_INTERPOLATION_KEY,
			undefined
		);
	});

	it('skips the checks named by ignoreChecks', () => {
		const input = { locales: { en: { a: 'Same' }, ko: { a: 'Same' } } };

		assert.strictEqual(
			analyzeTranslations(input, { target: 'en' }).issuesByCode.NOT_TRANSLATED_VALUE?.length,
			1
		);
		assert.strictEqual(
			analyzeTranslations(input, {
				target: 'en',
				ignoreChecks: 'NOT_TRANSLATED_VALUE'
			}).issuesByCode.NOT_TRANSLATED_VALUE,
			undefined
		);
	});

	it('reports an unusable locale as an invalid file', () => {
		const result = analyzeTranslations(
			{ locales: { en: { a: 'A' }, ko: null as any } },
			{ target: 'en' }
		);

		assert.strictEqual(result.issuesByCode.INVALID_FILE?.length, 1);
	});
});

describe('createAnalyzer().checkEntry', () => {
	const analyzer = createAnalyzer({ target: 'en' });

	it('checks a single key across locales', () => {
		const issues = analyzer.checkEntry({
			key: 'greeting',
			values: { en: 'Hello {name}', ko: '안녕하세요' }
		});

		assert.strictEqual(issues.length, 1);
		assert.strictEqual(issues[0].code, CHECK_CODE.NO_INTERPOLATION_KEY);
		assert.strictEqual(issues[0].interpolation, 'name');
		assert.strictEqual(issues[0].key, 'greeting');
	});

	it('returns nothing for a key with no problem', () => {
		assert.deepStrictEqual(
			analyzer.checkEntry({ key: 'a', values: { en: 'Hello', ko: '안녕' } }),
			[]
		);
	});

	it('reports a missing locale when told which locales exist', () => {
		const issues = analyzer.checkEntry({
			key: 'a',
			values: { en: 'Hello' },
			locales: ['en', 'ko']
		});

		assert.strictEqual(issues.length, 1);
		assert.strictEqual(issues[0].code, CHECK_CODE.NO_KEY);
		assert.strictEqual(issues[0].locale, 'ko');
	});

	it('never reports checks that need more than one key', () => {
		const issues = analyzer.checkEntry({
			key: 'a',
			values: { en: 'Same', ko: 'Same' }
		});

		assert.ok(issues.every((issue) => issue.code !== CHECK_CODE.DUPLICATE_VALUE));
	});

	it('carries the group through to the issue', () => {
		const issues = analyzer.checkEntry({
			key: 'a',
			values: { en: 'Hello', ko: '' },
			group: 'common.json'
		});

		assert.strictEqual(issues[0].group, 'common.json');
	});

	it('agrees with a full analysis of the same data', () => {
		const locales: Record<string, Record<string, string>> = {
			en: { a: 'Hello {name}', b: 'Bye' },
			ko: { a: '안녕하세요', b: '' }
		};
		const analyzed = analyzeTranslations({ locales }, { target: 'en', flattened: true });
		const incremental = Object.keys(locales.en).flatMap((key) =>
			analyzer.checkEntry({ key, values: { en: locales.en[key], ko: locales.ko[key] } })
		);

		assert.deepStrictEqual(
			incremental.map((issue) => `${issue.code}:${issue.key}`),
			analyzed.issues.map((issue) => `${issue.code}:${issue.key}`)
		);
	});
});

File: packages/dart/test/analyze_test.dart
import 'package:chki18n/chki18n.dart';
import 'package:test/test.dart';

void main() {
  group('analyzeTranslations', () {
    test('flattens nested translations before comparing them', () {
      final result = analyzeTranslations(
        const Chki18nInput(
          locales: {
            'en': {
              'desc': {'hello': 'Hello', 'bye': 'Goodbye'},
            },
            'ko': {
              'desc': {'hello': '안녕하세요'},
            },
          },
        ),
        options: const Chki18nOptions(target: 'en'),
      );

      expect(result.keyCount, 2);
      expect(result.success, isFalse);
      expect(result.of(Chki18nCheckCode.noKey).first.key, 'desc.bye');
    });

    test('uses already flattened translations as they are', () {
      final result = analyzeTranslations(
        const Chki18nInput(
          locales: {
            'en': {'a.b': 'Hi'},
            'ko': {'a.b': ''},
          },
        ),
        options: const Chki18nOptions(target: 'en', flattened: true),
      );

      expect(result.keyCount, 1);
      expect(result.of(Chki18nCheckCode.emptyValue).length, 1);
    });

    test('does no file system work, so it reports no files', () {
      final result = analyzeTranslations(
        const Chki18nInput(
          locales: {
            'en': {'a': 'A'},
            'ko': {'a': 'ㄱ'},
          },
        ),
        options: const Chki18nOptions(target: 'en'),
      );

      expect(result.files, isEmpty);
      expect(result.fileFormat, isNull);
    });

    test('compares each group on its own', () {
      final result = analyzeTranslations(
        const Chki18nInput(
          groups: {
            'common.json': {
              'en': {'a': 'A'},
              'ko': {'a': 'ㄱ'},
            },
            'errors.json': {
              'en': {'b': 'B'},
              'ko': <String, Object?>{},
            },
          },
        ),
        options: const Chki18nOptions(target: 'en'),
      );

      expect(result.groups, ['common.json', 'errors.json']);
      expect(result.of(Chki18nCheckCode.noKey).length, 1);
      expect(result.of(Chki18nCheckCode.noKey).first.group, 'errors.json');
    });

    test('honours custom interpolation delimiters', () {
      const input = Chki18nInput(
        locales: {
          'en': {'a': 'Hello {{name}}'},
          'ko': {'a': '안녕하세요'},
        },
      );

      expect(
        analyzeTranslations(
          input,
          options: const Chki18nOptions(
            target: 'en',
            interpolationPrefix: '{{',
            interpolationSuffix: '}}',
          ),
        ).of(Chki18nCheckCode.noInterpolationKey).length,
        1,
      );

      // The default single-brace delimiters do not recognise `{{name}}`, so the
      // missing placeholder goes unnoticed.
      expect(
        analyzeTranslations(
          input,
          options: const Chki18nOptions(target: 'en'),
        ).issuesByCode[Chki18nCheckCode.noInterpolationKey],
        isNull,
      );
    });

    test('skips the checks named by ignoreChecks', () {
      const input = Chki18nInput(
        locales: {
          'en': {'a': 'Same'},
          'ko': {'a': 'Same'},
        },
      );

      expect(
        analyzeTranslations(
          input,
          options: const Chki18nOptions(target: 'en'),
        ).of(Chki18nCheckCode.notTranslatedValue).length,
        1,
      );
      expect(
        analyzeTranslations(
          input,
          options: const Chki18nOptions(
            target: 'en',
            ignoreChecks: [Chki18nCheckCode.notTranslatedValue],
          ),
        ).issuesByCode[Chki18nCheckCode.notTranslatedValue],
        isNull,
      );
    });
  });

  group('createAnalyzer().checkEntry', () {
    final analyzer = createAnalyzer(options: const Chki18nOptions(target: 'en'));

    test('checks a single key across locales', () {
      final issues = analyzer.checkEntry(
        const Chki18nEntry(key: 'greeting', values: {'en': 'Hello {name}', 'ko': '안녕하세요'}),
      );

      expect(issues.length, 1);
      expect(issues.first.code, Chki18nCheckCode.noInterpolationKey);
      expect(issues.first.interpolation, 'name');
      expect(issues.first.key, 'greeting');
    });

    test('returns nothing for a key with no problem', () {
      expect(
        analyzer.checkEntry(const Chki18nEntry(key: 'a', values: {'en': 'Hello', 'ko': '안녕'})),
        isEmpty,
      );
    });

    test('reports a missing locale when told which locales exist', () {
      final issues = analyzer.checkEntry(
        const Chki18nEntry(key: 'a', values: {'en': 'Hello'}, locales: ['en', 'ko']),
      );

      expect(issues.length, 1);
      expect(issues.first.code, Chki18nCheckCode.noKey);
      expect(issues.first.locale, 'ko');
    });

    test('never reports checks that need more than one key', () {
      final issues = analyzer.checkEntry(
        const Chki18nEntry(key: 'a', values: {'en': 'Same', 'ko': 'Same'}),
      );

      expect(issues.every((issue) => issue.code != Chki18nCheckCode.duplicateValue), isTrue);
    });

    test('carries the group through to the issue', () {
      final issues = analyzer.checkEntry(
        const Chki18nEntry(key: 'a', values: {'en': 'Hello', 'ko': ''}, group: 'common.json'),
      );

      expect(issues.first.group, 'common.json');
    });

    test('agrees with a full analysis of the same data', () {
      const locales = <String, TranslationMap>{
        'en': {'a': 'Hello {name}', 'b': 'Bye'},
        'ko': {'a': '안녕하세요', 'b': ''},
      };
      final analyzed = analyzeTranslations(
        const Chki18nInput(locales: locales),
        options: const Chki18nOptions(target: 'en', flattened: true),
      );
      final incremental = [
        for (final key in locales['en']!.keys)
          ...analyzer.checkEntry(
            Chki18nEntry(key: key, values: {'en': locales['en']![key], 'ko': locales['ko']![key]}),
          ),
      ];

      expect(
        [for (final issue in incremental) '${issue.code.code}:${issue.key}'],
        [for (final issue in analyzed.issues) '${issue.code.code}:${issue.key}'],
      );
    });
  });
}
